### PR TITLE
Allow for sub RDF Sources via contains

### PR DESCRIPTION
**Pull request by terrellt**
I'll squash this when it's okay - for now it shows a refactor route.

### CODE REVIEW

**Comment by terrellt** [comment]
A side-note - this may get us close to a fix for #858

File: lib/active_fedora/association_hash.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
+require 'forwardable'
+
+module ActiveFedora
+  class AssociationHash
+    extend Forwardable
```
Comment (jeremyf) [inline lib/active_fedora/association_hash.rb]: Why does this extend Forwardable?

Comment (terrellt) [inline lib/active_fedora/association_hash.rb]: No idea. As far as I can tell it doesn't use it - was brought over from FilesHash. I could remove it, but I'm not sure what the project's policy on refactors + feature requests in one PR is.

Comment (mjgiarlo) [inline lib/active_fedora/association_hash.rb]: Unless @jcoyne voices an objection, I'd say dump it. It's a new file, so meh.

Comment (jcoyne) [inline lib/active_fedora/association_hash.rb]: Yes, dump it.

File: lib/active_fedora/associations/basic_contains_association.rb [hunk 2]
```diff
@@ -7,13 +7,21 @@ def reader(force_reload = false)
       end
 
       def find_target
-        reflection.build_association(target_uri) do |record|
+        find_or_initialize_target do |record|
           configure_datastream(record) if reflection.options[:block]
         end
       end
 
+      def find_or_initialize_target(&block)
```
Comment (jcoyne) [inline lib/active_fedora/associations/basic_contains_association.rb]: Should this be private?

File: lib/active_fedora/persistence.rb [hunk 3]
```diff
@@ -146,7 +146,9 @@ def create_record(options = {})
       serialize_attached_files
       @ldp_source = @ldp_source.create
       assign_uri_to_attached_files
+      assign_uri_to_child_resources
```
Comment (jcoyne) [inline lib/active_fedora/persistence.rb]: Is a child resource a hash-code resource?

**Comment by jcoyne** [comment]
Do we really want to make a clear distinction between attached files and child_resources? I'd prefer we handle them in the same way if possible.

File: lib/active_fedora/persistence.rb (hunk 3 above)
Comment (terrellt) [inline lib/active_fedora/persistence.rb]: No. Child resources are like /obj1 is an RDFSource and /obj1/obj2 is also a RDFSource. "contained_rdf_source" may be more appropriate.

**Comment by terrellt** [comment]
@jcoyne Can you take another look? Did some work to combine handling files and nested rdf sources.

**Comment by terrellt** [comment]
Rebased.

File: lib/active_fedora/associations/id_composite.rb [hunk 4]
```diff
@@ -20,7 +20,7 @@ def each
     private
 
     def convert(id)
-      if id.start_with?("http")
```
Comment (jcoyne) [inline lib/active_fedora/associations/id_composite.rb]: Please add YARD documentation on the parameter please

File: lib/active_fedora/core.rb [hunk 5]
```diff
@@ -37,6 +37,12 @@ def initialize(attributes_or_id = nil, &block)
       run_callbacks :initialize
     end
 
+    def uri=(uri)
```
Comment (jcoyne) [inline lib/active_fedora/core.rb]: Please add YARD doc about the parameter type(s)

File: lib/active_fedora/core.rb [hunk 6]
```diff
@@ -37,6 +37,12 @@ def initialize(attributes_or_id = nil, &block)
...
+      unless persisted?
```
Comment (jcoyne) [inline lib/active_fedora/core.rb]: Perhaps raise an error when calling this on a persisted object.

File: lib/active_fedora/associations/id_composite.rb (hunk 4 above)
Comment (terrellt) [inline lib/active_fedora/associations/id_composite.rb]: For the private method?

File: lib/active_fedora/core.rb (hunk 6 above)
Comment (terrellt) [inline lib/active_fedora/core.rb]: Recommendation on the exception class?

**Pull request merged**